pyopencl/tools.py

```diff
@@ -547,7 +554,20 @@ def match_dtype_to_c_struct(device, name, dtype, context=None):
     c_fields = []
     for field_name, dtype_and_offset in fields:
         field_dtype, offset = dtype_and_offset[:2]
-        c_fields.append("  %s %s;" % (dtype_to_ctype(field_dtype), field_name))
+        if hasattr(field_dtype, "subdtype") and field_dtype.subdtype is not None:
+            array_dtype = field_dtype.subdtype[0]
+            array_dims = field_dtype.subdtype[1]
+            dims_str = ""
+            try:
+                for dim in array_dims:
+                    dims_str += "[%d]" % dim
+            except TypeError:
+                dims_str = "[%d]" % array_dims
+            c_fields.append("  %s %s%s;" % (
+                dtype_to_ctype(array_dtype), field_name, dims_str)
+            )
```

Comment:
If the `subdtype` is another array, will this compose the declarator correctly? If not, check for unhandled cases and raise `NotImplementedError` in those.

Reply:
raised a `NotImplementedError` (sidenote: nested array dtypes are meaningless in my opinion, if the user wants to define a multidimensional array they can do it in one level, i.e. the way I have done it in the test I added)